proposals/lateinit-property-intrinsics.md

```diff
@@ -0,0 +1,224 @@
+
+We'd like to allow calling `isInitialized` and `reset` on as many properties and from as many different contexts as possible. However, without certain limitations we would have problems with source and binary compatibility.
+
+The main limitation is that we do not allow calling `isInitialized` or `reset` on a lateinit property declared in another class, lexically unrelated to the class or file of the call site. It's not that we don't know what bytecode to generate. If the property (and thus its backing field) is public, we could generate `GETFIELD`/`PUTFIELD` exactly in the same way as for the property in the containing class. However, doing so would make removing the `lateinit` modifier on a property, which is today merely an implementation detail, a source-breaking and binary-breaking change. We don't want to make the fact that the property is `lateinit` a part of its API, so we disallow usages from other classes.
+
+``` kotlin
+class Test {
+    lateinit var file: File
+}
+
+class Other {
+    fun test() {
+        Test()::file.isInitialized    // Error!
+    }
+}
+```
+
+Note that we could allow usages from other classes _in the same file_ but for simplicity, and to avoid errors when moving classes between files, we disallow such usages.
+
+Note that we _do_ allow usages from nested/inner classes of the property's declaring class, as well as from any lambdas and local classes inside members of the declaring class.
+
+``` kotlin
+class Test {
+    lateinit var file: File
+
+    inner class Inner {
+        fun test(t: Test) {
+            this@Test::file.isInitialized    // OK
+            run {
+                t::file.isInitialized        // OK
+            }
+        }
+    }
+}
+```
+
+In case the property is private and is accessed from a nested/local class or lambda, the JVM back-end must generate a synthetic accessor for the backing field:
+
+``` kotlin
+class Test {
+    private lateinit var file: File
+
+    fun test() {
+        run {
+            this::file.isInitialized    // ALOAD 0, INVOKESTATIC Test.access$getFile, IFNULL ...
+        }
+    }
+}
+```
+
+## Tooling support
+
+Code completion in the IDE must only propose `isInitialized` and `reset` if the calls are going to be allowed by the compiler, i.e. on an accessible lateinit property references.
+
+## Known issues
+
+* The solution is admittedly very ad-hoc. The `AccessibleLateinitPropertyLiteral` annotation serves only one particular purpose which is hardly generalizable to be useful to solve any other problems.
+* Because of the weird nature of the proposed call checker, trivial refactorings like "Extract variable" or "Use .let" could break the code here in an unexpected way:
+    ``` kotlin
+    if (this::file.isInitialized) ...             // OK
+
+    // Extract variable
+
+    val property = this::file
+    if (property.isInitialized) ...               // Error!
+
+    // Use .let
+
+    if (this::file.let { it.isInitialized }) ...  // Error!
```

Comment:
Also, looks like you can't `reset()` several `lateinit var`s at once:
`listOf(this::file, this::environment, this::other).all { it.reset() }`

Unlike `.let {}`, this scenario seems to me realistic.